salt/modules/junos.py

```diff
@@ -92,277 +101,634 @@ def facts():
+    if 'ttl' in op:
+        op['ttl'] = str(op['ttl'])
+
+    ret['out'] = True
+    try:
+        ret['message'] = jxmlease.parse(etree.tostring(conn.rpc.ping(**op)))
+    except Exception as exception:
+        ret['message'] = 'Execution failed due to "{0}"'.format(exception)
         ret['out'] = False
     return ret
 
 
-def cli(command=None):
+def cli(command=None, format='text', **kwargs):
     '''
-    Executes the CLI commands and reuturns the text output.
+    Executes the CLI commands and returns the output in specified format. \
+    (default is text) The ouput can also be stored in a file.
 
     Usage:
 
     .. code-block:: bash
 
-        salt 'device_name' junos.cli 'show version'
+        salt 'device_name' junos.cli 'show system commit'
 
+        salt 'device_name' junos.cli 'show version' dev_timeout=40
+
+        salt 'device_name' junos.cli 'show system alarms' 'xml' dest=/home/user/cli_output.txt
+
+
+    Parameters:
+      Required
+        * command:
+          The command that need to be executed on Junos CLI. (default = None)
+      Optional
+        * format:
+          Format in which to get the CLI output. (text or xml, \
+            default = 'text')
+        * kwargs: Keyworded arguments which can be provided like-
+            * dev_timeout:
+              Set NETCONF RPC timeout. Can be used for commands which
+              take a while to execute. (default = 30 seconds)
+            * dest:
+              The destination file where the CLI output can be stored.\
+               (default = None)
 
-    Options:
-      * command: The command that need to be executed on Junos CLI.
     '''
     conn = __proxy__['junos.conn']()
+
+    # Cases like salt 'device_name' junos.cli 'show system alarms' ''
+    # In this case the format becomes '' (empty string). And reply is sent in xml
+    # We want the format to default to text.
+    if not format:
+        format = 'text'
+
     ret = dict()
-    ret['message'] = conn.cli(command)
+    if command is None:
+        ret['message'] = 'Please provide the CLI command to be executed.'
+        ret['out'] = False
+        return ret
+
+    op = dict()
+    if '__pub_arg' in kwargs:
+        if kwargs['__pub_arg']:
+            if isinstance(kwargs['__pub_arg'][-1], dict):
+                op.update(kwargs['__pub_arg'][-1])
+    else:
+        op.update(kwargs)
+
+    try:
+        result = conn.cli(command, format, warning=False)
+    except Exception as exception:
+        ret['message'] = 'Execution failed due to "{0}"'.format(exception)
+        ret['out'] = False
+        return ret
+
+    if format == 'text':
+        ret['message'] = result
+    else:
+        result = etree.tostring(result)
+        ret['message'] = jxmlease.parse(result)
+
+    if 'dest' in op and op['dest'] is not None:
+        with fopen(op['dest'], 'w') as fp:
+            fp.write(result)
+
     ret['out'] = True
     return ret
 
 
-def shutdown(time=0):
+def shutdown(**kwargs):
```

Comment:
With our ansible experience, we see reboot being used many times after software installation. But not sure of shutdown :)